src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -282,6 +282,16 @@ private static string GenerateUpgradeReport(string projectName, string backupPat
             }
         }
 
+        private static void AddToExisting(IPackageReferenceContextInfo pkg, ISet<Tuple<string, string>> existingPackages)
+        {
+            PackageIdentity package = pkg.Identity;
+            Tuple<string, string> packageInfo = Tuple.Create(package.Id, package.Version == null ? "" : package.Version.ToNormalizedString());
+            if (!existingPackages.Contains(packageInfo))
```

Comment:
Do we need to override the hash of Tuple? Not sure how it works for Tuple.

Reply:
Also do we really need to test if set contains the tuple? Why don't just try to add it? Since it's set it only accept 1 copy of same tuple, right?